### ISSUE TITLE
.NET Core 1 support

### ISSUE THREAD

**Issue by 206mph** [rexm/Handlebars.Net#116]
Any chance Handlebars.Net will support .Net Core in the very near future?  Or, maybe it does and I'm just missing something.

**Comment by rexm** [comment]
I haven't tested it on .Net Core yet, so I don't know how close or far it is from being compatible. I welcome any initial research to help move it along.

**Comment by 206mph** [comment]
I've been working on getting a version of this working on my machine.  I've got most of the errors resolved but am stumbling on one error that I just am not familiar with and is occurring in several files.  For example, in Compiler > Translation > Expression > BlockHelperFunctionBinder.cs, there are a couple lines of code that read helper.Method.  This is throwing an error because there is no definition for 'Method' and no extension method for 'Method'.

It appears this is coming from System.Delegate in the current version.  That's just my guess as some of this is a little beyond what this kind of work I do in C#.  Any info would be great on where it's coming from or what it does.  If I get that issue resolved, I will be down to 1 bug before I can compile/test.

**Comment by rexm** [comment]
I believe there are some changes to how Core handles reflection

**Comment by sandorfr** [comment]
Hi,

Indeed, there are some changes in the .net core reflection but it's basically the same as it was with portable libraries.

I opened a pull request supporting netstandard 1.5 (I think it's possible to actually target a lower standard, I'll do some tests later). It also targets net451 (so the dotnet pack command generates a package fully compatible with any .net core app).

I chose to reference existing source code instead of duplicating it. I had to add netstandard compilation variable. I think this should not create too much maintainability issue. 

Looking forward to your feedback on how to improve it and make it mergeable :). Great work btw, I was using mustache-sharp so far and I love the "true handlebars" approach.

**Comment by rexm** [comment]
Awesome start. Made a quick pass through this morning; will review in more detail this evening, but looks pretty straightforward.

I didn't see an update to the nuspec to target the correct framework - is that an oversight?

**Comment by sandorfr** [comment]
Your're right I did not update the nuspec. I'll look into this tonight.

However, since `project.json` acts as a sort of nuspec file, you can generate one easily running `dotnet pack -c Release --version-suffix dotnet-core-rc2-0001` in the `project.json` directory. (I have a packaged one available there : https://www.myget.org/feed/sandor-public/package/nuget/Handlebars.NetStandard)

**Comment by sandorfr** [comment]
just updated the nuspec file. I think you should not merge it to your master branch until dotnet core is RTM and publish package with a prerelease tag until then (just my 2 cents).

**Comment by sandorfr** [comment]
Worth noting that a side effect of netstandard is that windows 10 is supported (#102)

**Comment by sandorfr** [comment]
.net core RTM preview package available here : https://www.nuget.org/packages/Handlebars.NetStandard/1.0.0-dotnet-core-rtm1

**Comment by EvilBeaver** [comment]
@sandorfr excuse me, does it still depends on Framework 4.5 anyway? nuget.org shows old .NET in the dependency list.

**Comment by rexm** [comment]
Sorry, I should have closed this issue. The official release does now support dotnet core and .NET 4.x. Both versions are in the same package.

**Comment by priedthirdeye** [comment]
Can someone point me in the right direction for using Handlebars.Net as the view engine for Asp.Net core?

I'm coming from node.js development and I'm having a hard time understanding how to use Handlebars.net instead of the default Razor view engine.

**Comment by rexm** [comment]
The Handlebars library doesn't have turnkey support for ASP.NET view engines. It requires some framework-level work on your part to get it all working.

**Comment by sandorfr** [comment]
@priedthirdeye 

Just my two cents, but as much as I love handlebars, I would advise against using it as your view engine. 

* it's far less powerful as Razor (think about tag helpers for instance)
* you will be likely to be very isolated in your scenario, hence won't have community / microsoft support
* you will add a learning curve for any already asp.net fluent developer
* you will be on the bleeding edge

Kind regards.

**Comment by priedthirdeye** [comment]
So when Handlebars.Net was branched to "support .net core", was that not related to using it as a view engine for asp.net core?

Please note that I'm new to this technology, so my terminology and understanding may be off.

-----------

I have a use case where it may make sense to use handlebars as the view engine. I at least want to consider it before presenting options to my boss. 

Some Details:

We currently have a website that has several thousand templates using a custom template language very similar to handlebars. The architecture, written in c++ many years ago, passes input to stored procedures and prepares the results as variables and/or loops for the use in the templates. Employees, many of which are not "developers", are familiar with the template language and very basic logic structure. It is by design that we have a relatively limited amount of functionality at the template level. The owner wanted templates to be easy to understand for someone with basic HTML knowlege and basic training. Since I'm presenting alternatives to our existing system, I'd like to try to stay as close to his vision as possible. In my current direction using .NET core, I won't be using the "M" in the "MVC" architecture. I just want to create something that can execute our existing stored procedures, and make the data available to the template. I understand that I will lose out on a lot of benefits of .net core in doing it this way, but I get the impression from my preliminary review that true asp.net core MVC with the Razor view engine would likely add huge cost and time to the conversion effort and likely fuel a decision to just keep things the way they are. We are not building from the ground up.

I already created a pretty mature proof of concept in node.js, Express and Handlebars that does what I want. Right before I was preparing to present a proposal for new architecture I read an article on asp.net core being faster than node.js. That's why I'm trying to create asp.net core proof of concepts as well.

**Comment by priedthirdeye** [comment]
The documentation for a custom view engine is just a placeholder:

https://docs.microsoft.com/en-us/aspnet/core/mvc/views/custom-view-engine

And it appears people aren't having much luck implementing a custom view engine:

https://github.com/aspnet/Docs/issues/137

**Comment by priedthirdeye** [comment]
I didn't see the comment from @rexm when I responded with my question. 

Thanks for your answers.

**Comment by sandorfr** [comment]
You're right this is just support for .netstandard 1.5 which means that you can reference this library from a net core app. 

Support as a view engine is another subject which could be brought as an additional library on top of this one.

**Comment by rexm** [comment]
I agree with @sandorfr... Handlebars is likely a very good candidate for your use case with such a large set of templates, but I would recommend building something simpler and more specific to your exact use case _on top of_ ASP.NET (or any mainstream framework/platform) instead of trying to repurpose the very sophisticated plumbing of the framework itself.

**Comment by priedthirdeye** [comment]
Thanks. I now understand that using asp.net core MVC is overkill for what I am trying to accomplish. I'm now moving forward with a non-MVC asp.net core implementation.

Do you know of a boilerplate available for a .net core project that just utilizes .net core (no MVC), has routing and serving templates (ideally using Handlebars.net already)? I looked and couldn't find anything. 

I just started a regular asp.net core application without MVC, added routing and I am now working on figuring out how to serve templates and ultimately utilize Handlebars.Net. I'm hoping there is a boilerplate that will get me further along from the start.

**Comment by rexm** [comment]
I'm puzzled by your approach. You said:
>Right before I was preparing to present a proposal for new architecture I read an article on asp.net core being faster than node.js.

You decided to pivot away from a technology that you're clearly familiar enough with to build a working prototype because of an article? That's really puzzling. Both nodejs and asp.net can be highly performant, far beyond the typical needs for all but the heaviest of loads. The vast majority of whether an app can handle the performance requirements of it fall to the implementation of the app, _not_ the framework the app is running on. All frameworks at their base are fast enough.

I would strongly recommend you stick with what you already know works.

**Comment by priedthirdeye** [comment]
I want to present available options along with the pros and cons of each so my boss can make an informed decision. I am personally partial to NodeJS and hope he goes that direction. He comes from a C++ background and may object to transitioning to NodeJS.  If the same thing I accomplished with NodeJS can be accomplished in a practical manner using .net core/c#/handlebars, I suspect that he would go that direction. 

I'm just trying to do a proof of concept, not a full implementation.

------

So my current focus is using asp.net core, loading templates into cache, and rendering using Handlebars.Net library. It isn't clear to me if this is breaking new ground. Either way I am exploring the feasibility of .net core and will present the outcome to him regardless of the outcome.

**Comment by priedthirdeye** [comment]
I'm making pretty good progress.

I got routing, basic template caching and partials working.

Can you please provide me with some insight on how "layout.hbs" is supposed to work?

I don't see how Handlebars.Net knows to associated the layout.hbs with the pages and there isn't much in the documentation regarding layout.hbs. 

This is my work in progress code ("Handlebars.Net-core-boilerplate"). I included comments in the file.

https://github.com/priedthirdeye/Handlebars.Net-core-boilerplate/blob/master/Startup.cs

**Comment by sandorfr** [comment]
I have a love/hate relationship with nodejs, so I would be biased and won't jump into that part of the discussion. I'm happy to see people coming from non microsoft background considering aspnet core.

I'm not sure to understand what you mean by `layout.hbs`. What I know for sure, is that you could leverage a bit more of MVC writing an `HandlerbarsResult` (which would take an object and a template and output html).  Actions Results are objects a controller action is expected to return which will be in charge of producing the output, (old asp.net doc is still valid https://msdn.microsoft.com/en-us/library/system.web.mvc.actionresult(v=vs.118).aspx). 

I have a sample on extending objectResult [here](https://github.com/geeklearningio/gl-dotnet-domain/blob/develop/src/GeekLearning.Domain.AspnetCore/MaybeResult.cs). But in your case as you don't need the formatters, I would rather juste implement `IActionResult`.

**Comment by sandorfr** [comment]
To address your caching issue you can rely on `IMemoryCache` as simple and efficient as your Dictionnary, `GetOrCreate` is a treat.

**Comment by rexm** [comment]
What is layout.hbs?

**Comment by priedthirdeye** [comment]
Thanks. I will absorb and apply that feedback.

Regarding my `layout.hbs` confusion, one of the handlebars libraries I'm accustomed to describes the layout file(s) as follows

"The main layout is the HTML page wrapper which can be reused for the different views of the app. {{{body}}} is used as a placeholder for where the main content should be rendered."

It's basically like "master pages" were in old asp.net. 

So when I see `Views\layout.hbs` in the Handebars.Net documentation, I assumed there was some functionality baked in that will allow layout.hbs to function as a "master page" with a placeholder

Here's an example of what a layout.hbs file might look like (taken from an example in one of the libraries):

https://github.com/pillarjs/hbs/blob/master/examples/partial/views/layout.hbs

Note the {{{body}}} which is replaced with the contents of the page being requested.

Does this functionality exist at all in Handlebars.Net?

**Comment by rexm** [comment]
Handlebars.Net is not a view engine. That functionality can be built using Handlebars.Net, but it does not have any concept baked-in of view engines.

**Comment by priedthirdeye** [comment]
@sandorfr 

I supported the concept of layouts in my boilerplate.

Unfortunately I was unsuccessful with both attempts to apply your feedback on using IMemoryCache and creating a HandlebarsResult class that inherits from the "ActionResult" class (not sure my terminology is correct).

I'm wondering if you will be will to contribute to my boilerplate project and I will pay you for your time. I'd like to keep it at 2 billing hours or less though. I just added my email address to my profile if you are interested.

I'm hoping two hours would be enough to pick up where I left off and complete the following requirements

1) Support IMemoryCache to cache templates when in production. Pull from disk when env.IsDevelopment()
2) Support the concept of controllers to handle the logic associated with a route before the data is passed to a view.
3) Support an Array/List in the data structure passed to the template so {{#each}} can be used.

Let me know. Thanks.

https://github.com/priedthirdeye/Handlebars.Net-core-boilerplate

**Comment by sandorfr** [comment]
Ok, let met take a closer look and I'll contact you by email.

**Comment by priedthirdeye** [comment]
@sandorfr

Is this something you are interested in doing before the end of the week. Is so please email me with your rate.

**Comment by sandorfr** [comment]
For the record of anyone willing to go beyond to what's done in the boilerplate regarding aspnet core integration, building a viewengine should be qui straightforward. 

At the moment there is an effort regarding a Markdown ViewEngine which might be used as a reference / inspiration:
* [Introduction Blog post](http://www.hishambinateya.com/introducing-a-new-markdown-view-engine-for-asp.net-core)
* [Github project](https://github.com/hishamco/MarkdownViewEngine)

I think this should be way more easier with handlebars as it has almost everything required in it's core as opposed to markdown which requires a lot of extensions.

**Comment by kayweng** [comment]
did the handlerbar.Net already support dotnet core? It seems didn't work for my environment which is dotnet core 1.0 ...

**Comment by sandorfr** [comment]
@kayweng yes it does, thanks to netstandard 1.4 target as you can see on nuget : https://www.nuget.org/packages/Handlebars.Net/ . If you have an issue, you should provide a bit more details.

**Comment by rexm** [comment]
@kayweng this is not a catchall for any dotnet core / standard issue. Please open a new issue with the details of your specific problem.

**Comment by kayweng** [comment]
thanks for the prompt reply. I found my error issue that is AWS lambda only support dotnet core 1.0 and handlerbars.net is using a higher version(1.1?). Never mind, I will do a custom change to fit my scenario.
